### PR TITLE
[Wav2Vec2] Fix tokenizer set lang

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
# What does this PR do?

The PR #23909 removed `unique_no_split_tokens` as an attribute of the Wav2Vec2 tokenizer, however it is required to set the language in the method `.set_target_lang`:

https://github.com/huggingface/transformers/blob/dcbfd93d7aeb14f8ff08a48866d2a68950d4c69a/src/transformers/models/wav2vec2/tokenization_wav2vec2.py#L237

Thus, calling `.set_target_lang` currently throws an error:

```python
from transformers import Wav2Vec2CTCTokenizer

tokenizer = Wav2Vec2CTCTokenizer.from_pretrained("facebook/mms-1b-all")
tokenizer.set_target_lang("spa")
```

**Output:**
```
Traceback (most recent call last):
  File "/Users/sanchitgandhi/transformers/debug_tokenizer.py", line 4, in <module>
    tokenizer.set_target_lang("spa")
  File "/Users/sanchitgandhi/transformers/src/transformers/models/wav2vec2/tokenization_wav2vec2.py", line 237, in set_target_lang
    self.unique_no_split_tokens.append(token)
AttributeError: 'Wav2Vec2CTCTokenizer' object has no attribute 'unique_no_split_tokens'
```

This PR re-instates `unique_no_split_tokens` as an attribute of the tokenizer. Note that this should already be tested for in the following test: https://github.com/huggingface/transformers/blob/dcbfd93d7aeb14f8ff08a48866d2a68950d4c69a/tests/models/wav2vec2/test_tokenization_wav2vec2.py#L798

### CODE REVIEW

**Comment by andergisomon** [comment]
My space suddenly had a runtime error because of this.

**Comment by sanchit-gandhi** [comment]
Sorry to hear that @andergisomon - note that running `transformers` on the latest PyPi version should bypass this error. It's only on `main` at the moment:
```
pip uninstall transformers -y
pip install --upgrade transformers
```

**Review by ArthurZucker — COMMENTED** [review]
Thanks, I'd rather remove the addition of these tokens to `unique_no_split_tokens` as they won't be used.( #26322 was on it's way)

The issue is that adding them to `self.unique_no_split_tokens` just does nothing. A fix that would keep the previous behaviour is to rather do `self.add_tokens(self.encoder.keys())`. It's a bit inefficient but will make sure they are not split!

**Comment by sanchit-gandhi** [comment]
Thanks for the review! Applied the suggestions from your comment - look good to you?

File: src/transformers/models/wav2vec2/tokenization_wav2vec2.py [hunk 1]
```diff
@@ -234,7 +229,7 @@ def set_target_lang(self, target_lang: str):
         # characters are not split at tokenization
         for token in self.encoder.keys():
             if len(token) > 1:
-                self.unique_no_split_tokens.append(token)
+                self.add_tokens(token)
```
Comment (ArthurZucker) [inline src/transformers/models/wav2vec2/tokenization_wav2vec2.py:232]: `add_tokens` loops over the tokens, so I'd recommend adding them using:
```suggestion
        self.add_tokens(self.encoder.keys())
```
as an internal checks makes sure `let(token)>1`

File: src/transformers/models/wav2vec2/tokenization_wav2vec2.py [hunk 2]
```diff
@@ -206,14 +206,9 @@ def __init__(
...
         # characters are not split at tokenization
-
-        # TODO @ArthurZ add them or just update the trie?
-        unique_no_split_tokens = []
         for token in self.encoder.keys():
             if len(token) > 1:
-                unique_no_split_tokens.append(AddedToken(token, rstrip=True, lstrip=True, normalized=False))
-
-        self.add_tokens(unique_no_split_tokens)
+                self.add_tokens(AddedToken(token, rstrip=True, lstrip=True, normalized=False))
```
Comment (ArthurZucker) [inline src/transformers/models/wav2vec2/tokenization_wav2vec2.py:211]: same comment here

**Review by ArthurZucker — APPROVED** [review]
Let's simplify a bit and good to go

File: src/transformers/models/wav2vec2/tokenization_wav2vec2.py (hunk 1 above)
Comment (sanchit-gandhi) [reply src/transformers/models/wav2vec2/tokenization_wav2vec2.py]: But we need `lstrip=rstrip=True` here? So need to use `AddedToken(lstrip=True, rstrip=True)` as we had before?

**Review by ArthurZucker — APPROVED** [review]
Fine with me!

**Pull request merged**